### DISCUSSION TITLE
Question: How are the Hyperparameter generated?

### DISCUSSION THREAD

**Discussion by BradKML** [winedarksea/AutoTS#141] [Q&A]
TL;DR wanting to understand https://github.com/winedarksea/AutoTS/blob/master/autots/models/sklearn.py

How are ARIMA, ETS, and Unobserved Components converted into Auto-TS tools with hyperparameter selection?

- AutoARIMA already exists separately as https://alkaline-ml.com/pmdarima/modules/generated/pmdarima.arima.AutoARIMA.html
- AutoETS also exists in other libraries as https://www.sktime.org/en/latest/api_reference/auto_generated/sktime.forecasting.ets.AutoETS.html
- It is unknown if Unobserved Components has AutoTS functionality, and AutoTS might have a solution within the library

**Comment by winedarksea** [comment]
-Firstly the models you referenced aren't in sklearn.py but statsmodels.py
Short answer: [Statsmodels](https://www.statsmodels.org/stable/index.html) package, I use what is defined there, which are also used by 'most' time series packages in Python. I add a wrapper that includes a parameter generation function.

There's also a new KalmanStateSpace model (in dev) which allows a structural state space definition of 'any' type, which all of the models you mentioned can be defined as, and will be the fastest implementation of those models in AutoTS.

**Reply by BradKML** [reply]
Are you referring to this one? https://www.statsmodels.org/dev/optimization.html
Also what is your view on Hyperopt or Optuna or Tune?

**Reply by winedarksea** [reply]
I have written my own genetic optimization that tunes the model parameters at the same time as the preprocessing. I believe auto-sklearn uses something similar, although I haven't explored their code.

There are two major problems with optimization in time series:
1. lots of discrete parameters while most optimization is for continuous space. For example, type of distance metric in motifs, or the averaging method for average value naive
2. lots of local minima on something like `seasonality` where most optimizers really struggle.

Likely you will need a combination of multiple optimization methods if you are building your own
